common/src/main/java/org/conscrypt/ConscryptEngine.java

```diff
@@ -622,6 +622,14 @@ private ConscryptSession provideHandshakeSession() {
         }
     }
 
+    // After handshake has started, provide active session otherwise a null session,
+    // for code which needs to read session attributes without triggering the handshake.
+    private ConscryptSession provideAfterHandshakeSession() {
+        return (state < STATE_HANDSHAKE_STARTED)
```

Comment:
do we have a test to assert that states are ordered according to handshake lifecycle? Otherwise we could introduce a new state that is > STATE_HANDSHAKE_STARTED but really is reached before STATE_HANDSHAKE_STARTED step.

Reply:
Yeah, I nearly added an extra state to handle this case but didn't to avoid exactly that issue.  Test added.